versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java

```
@@ -1684,4 +1709,129 @@ private void writeRefLog(Connection connection, AdapterTypes.RefLogEntry entry)
+    try {
+      try (PreparedStatement ps = conn.prepareStatement(SqlStatements.INSERT_ATTACHMENT)) {
+        attachments.forEach(
+            bc -> {
+              try {
+                attachmentInsert(ps, bc);
+              } catch (SQLException e) {
+                throw new RuntimeException(e);
+              }
+            });
+      }
+      txCommit(conn);
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      releaseConnection(conn);
+    }
+  }
+
+  private void attachmentInsert(PreparedStatement ps, ContentAttachment attachment)
+      throws SQLException {
+    ps.setString(1, config.getRepositoryId());
+    ps.setString(2, attachmentKeyAsString(ContentAttachmentKey.fromContent(attachment)));
+    ps.setBytes(3, ProtoSerialization.toProtoValue(attachment).toByteArray());
+    ps.setString(4, attachment.getVersion());
+    ps.executeUpdate();
+  }
+
+  @Override
+  public boolean consistentPutAttachment(
+      ContentAttachment attachment, Optional<String> expectedVersion) {
+    Connection conn = borrowConnection();
+    try {
+      if (expectedVersion.isPresent()) {
+        try (PreparedStatement ps = conn.prepareStatement(SqlStatements.UPDATE_ATTACHMENT)) {
+          ps.setBytes(1, ProtoSerialization.toProtoValue(attachment).toByteArray());
+          ps.setString(2, attachment.getVersion());
+          ps.setString(3, config.getRepositoryId());
+          ps.setString(4, attachmentKeyAsString(ContentAttachmentKey.fromContent(attachment)));
+          ps.setString(5, expectedVersion.get());
+          if (ps.executeUpdate() != 1) {
+            return false;
+          }
+        }
+      } else {
+        try (PreparedStatement ps = conn.prepareStatement(SqlStatements.INSERT_ATTACHMENT)) {
+          attachmentInsert(ps, attachment);
+        }
+      }
+      txCommit(conn);
+    } catch (SQLException e) {
+      if (isIntegrityConstraintViolation(e)) {
+        return false;
```

shouldn't we log that? Right now the information about the failure will be hidden (because we are just returning `false`)